### PR TITLE
Support chunk cache on a per-dataset basis

### PR DESCRIPTION
**Pull request by yuanqj8191**
The original API refers to:
https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-SetChunkCache

### CODE REVIEW

File: h5p_proplist_test.go [hunk 1]
```diff
@@ -168,3 +245,21 @@ func compare(ds0, ds1 []float64) error {
 	}
 	return nil
 }
+
+func checkChunkCahche(nslots, nbytes int, w0 float64, dapl *PropList) error {
```
Comment (sbinet) [inline h5p_proplist_test.go]: `s/Cahche/Cache/`

File: h5p_proplist_test.go [hunk 2]
```diff
@@ -168,3 +245,21 @@ func compare(ds0, ds1 []float64) error {
...
+	nslots_, nbytes_, w0_, err := dapl.GetChunkCache()
```
Comment (sbinet) [inline h5p_proplist_test.go]: not too fond of the `xxx_` convention.
perhaps postfix `Cache` instead? or something along these lines?

File: h5p_proplist_test.go [hunk 3]
```diff
@@ -132,14 +203,20 @@ func save(fn, dsn string, dims []uint, dclp *PropList) ([]float64, error) {
...
 	f, err := OpenFile(fn, F_ACC_RDONLY)
 	if err != nil {
 		return nil, err
 	}
 	defer f.Close()
 
-	dset, _ := f.OpenDataset(dsn)
+	var dset *Dataset
+	if dapl == nil {
+		dset, _ = f.OpenDataset(dsn)
```
Comment (sbinet) [inline h5p_proplist_test.go]: why do we ignore the returned error?

File: h5p_proplist.go [hunk 4]
```diff
@@ -95,6 +97,28 @@ func (p *PropList) SetDeflate(level int) error {
...
+// GetChunkCache retrieves the number of chunk slots in the raw data chunk cache hash table.
+// https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-GetChunkCache
+func (p *PropList) GetChunkCache() (nslots, nbytes int, w0 float64, err error) {
+	var (
+		c_nslots C.size_t
+		c_nbytes C.size_t
+		c_w0     C.double
+	)
+	if err = h5err(C.H5Pget_chunk_cache(C.hid_t(p.id), &c_nslots, &c_nbytes, &c_w0)); err != nil {
+		return
```
Comment (sbinet) [inline h5p_proplist.go]: please refrain from using naked returns.

File: h5p_proplist.go [hunk 5]
```diff
@@ -95,6 +97,28 @@ func (p *PropList) SetDeflate(level int) error {
...
+	}
+	nslots, nbytes, w0 = int(c_nslots), int(c_nbytes), float64(c_w0)
+	return
```
Comment (sbinet) [inline h5p_proplist.go]: please refrain from using naked returns.

File: h5p_proplist_test.go [hunk 6]
```diff
@@ -102,7 +102,78 @@ func TestDeflate(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	defer dapl.Close()
+
+	nslots, nbytes, w0, err := dapl.GetChunkCache()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	nslots_, nbytes_, w0_ := nslots*4, nbytes*2, w0/3
```
Comment (sbinet) [inline h5p_proplist_test.go]: not too fond of these `xyz_` variables.
could we find better names?

File: h5d_public.go [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package hdf5
+
+// #include "hdf5.h"
+import "C"
+
+const (
+	// Used to unset chunk cache configuration parameter.
```
Comment (sbinet) [inline h5d_public.go]: this should be moved to line 10, to comment the whole `const`-block.

File: h5d_public.go [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+	// https://support.hdfgroup.org/HDF5/doc/RM/RM_H5P.html#Property-SetChunkCache
+	D_CHUNK_CACHE_NSLOTS_DEFAULT int     = -1 // rdcc_nslots
```
Comment (sbinet) [inline h5d_public.go]: we could perhaps add a somewhat better comment, piggy-backing on the official documentation. (especially for `rdcc_w0`.)
(at the very least, we should explain what `rdcc` stands for.)

File: h5g_group.go [hunk 9]
```diff
@@ -108,6 +108,19 @@ func (g *CommonFG) OpenDataset(name string) (*Dataset, error) {
 	return newDataset(hid, nil), nil
 }
 
+// OpenDataset opens and returns a named Dataset with a user-defined PropList.
```
Comment (sbinet) [inline h5g_group.go]: `s/OpenDataset/OpenDatasetWith/`

File: h5p_proplist_test.go (hunk 1 above)
Comment (yuanqj8191) [reply h5p_proplist_test.go]: Fixed

File: h5p_proplist_test.go (hunk 2 above)
Comment (yuanqj8191) [reply h5p_proplist_test.go]: Fixed

File: h5p_proplist_test.go (hunk 3 above)
Comment (yuanqj8191) [reply h5p_proplist_test.go]: Fixed

File: h5p_proplist.go (hunk 4 above)
Comment (yuanqj8191) [reply h5p_proplist.go]: Fixed

File: h5p_proplist.go (hunk 5 above)
Comment (yuanqj8191) [reply h5p_proplist.go]: Fixed

File: h5p_proplist_test.go (hunk 6 above)
Comment (yuanqj8191) [reply h5p_proplist_test.go]: Fixed

File: h5d_public.go (hunk 8 above)
Comment (yuanqj8191) [reply h5d_public.go]: Fixed

File: h5g_group.go (hunk 9 above)
Comment (yuanqj8191) [reply h5g_group.go]: Fixed

File: h5p_proplist.go [hunk 10]
```diff
@@ -112,11 +112,8 @@ func (p *PropList) GetChunkCache() (nslots, nbytes int, w0 float64, err error) {
 		c_nbytes C.size_t
 		c_w0     C.double
 	)
-	if err = h5err(C.H5Pget_chunk_cache(C.hid_t(p.id), &c_nslots, &c_nbytes, &c_w0)); err != nil {
-		return
-	}
-	nslots, nbytes, w0 = int(c_nslots), int(c_nbytes), float64(c_w0)
-	return
+	err = h5err(C.H5Pget_chunk_cache(C.hid_t(p.id), &c_nslots, &c_nbytes, &c_w0))
+	return int(c_nslots), int(c_nbytes), float64(c_w0), nil
```
Comment (sbinet) [inline h5p_proplist.go]: `s/nil/err/`

Comment (yuanqj8191) [reply h5p_proplist.go]: Fixed

**Pull request merged**